agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1394 @@
+                    }
+                }
+                if remove_instance {
+                    trace!("update_connectivity_status - instance {} has been offline too long ... terminating device plugin", instance);
+                    device_plugin_service::terminate_device_plugin_service(
+                        &instance,
+                        self.instance_map.clone(),
+                    )
+                    .await
+                    .unwrap();
+                    k8s::try_delete_instance_arc(
+                        kube_interface.clone(),
+                        instance.clone(),
+                        self.config.metadata.namespace.as_ref().unwrap().clone(),
+                    )
+                    .await
+                    .unwrap();
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub mod start_discovery {
+    use super::super::registration::DiscoveryHandlerConnectivityStatus;
+    #[double]
+    pub use super::DiscoveryOperator;
+    use super::{try_receive, StreamType};
+    use akri_discovery_utils::discovery::DISCOVERY_HANDLER_PATH;
+    use akri_shared::k8s;
+    use mockall_double::double;
+    use std::{sync::Arc, time::Duration};
+    use tokio::sync::broadcast;
+
+    /// This is spawned as a task for each Configuration and continues to run
+    /// until the Configuration is deleted, at which point, this function is signaled to stop.
+    /// In a separate task, it calls connects to each discovery handler in the RegisteredDiscoveryHandlerMap
+    /// with the same protocol name as the Configuration (Configuration.protocol.name). Then, it listens for
+    /// updates from the discovery handler on what devices are currently visible to the node.
+    /// Passes this list to a function that updates the InstanceConnectivityStatus of the Configuration's Instances
+    /// or deletes Instance CRs if needed. If a new instance becomes visible that isn't in the Configuration's InstanceMap,
+    /// a DevicePluginService and Instance CRD are created for it, and it is added to the InstanceMap.
+    ///
+    /// It also spawns a task to check whether Offline Instances have exceeded their grace period, in which case it
+    /// deletes the Instance.
+    pub async fn start_discovery(
+        discovery_operator: DiscoveryOperator,
+        new_discovery_handler_sender: broadcast::Sender<String>,
+        stop_all_discovery_sender: broadcast::Sender<()>,
+        finished_all_discovery_sender: &mut broadcast::Sender<()>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        let config = discovery_operator.get_config();
+        info!(
+            "start_discovery - entered for protocol {}",
+            config.spec.protocol.name
+        );
+        let config_name = config.metadata.name.clone();
+        let mut tasks = Vec::new();
+        let discovery_operator = Arc::new(discovery_operator);
+        let task1_discovery_operator = discovery_operator.clone();
```

Comment:
I'm not clear on the difference between task1 and task2. It seems one is always only discovering while the other monitors for stopping as well?

Reply:
I updated the documentation on `start_discovery` to delineate what each task does in 80e02b7. Task1 and Task2 are comparable to how agent and controller check for pre-existing and new configurations and instances, respectively. Here, `start_discovery` is checking for already registered and newly registered discovery handlers. The second task has a channel to notify it to stop checking when the associated Configuration is deleted.